### DISCUSSION TITLE
How to perform a scroll to a element on android when using espresso?

### DISCUSSION THREAD

**Discussion by attrideepak** [appium/appium#16347] [Q&A] [answered]
I want to perform a scroll to an element and I am using appium espresso. How can I do it?

**Comment by KazuCocoa** [comment, marked as answer]
https://github.com/appium/appium-espresso-driver such as `mobile: scrollToPage` (via espresso' snavigateTo api) or https://appium.io/docs/en/commands/interactions/actions/ such as https://github.com/appium/ruby_lib_core/blob/1632637fd872c0b80dfb97b8514ada6a7164eebf/test/functional/common_w3c_actions.rb#L16 (general gesture api) can do it.
Asking questions in https://discuss.appium.io/ also may help

**Reply by attrideepak** [reply]
I get the following error when I use data matcher.
*Original error: androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: is assignable from class: class android.widget.AdapterView*

I also tried using mobile:scrollToPage but it does not work. I am getting the following error in this case.

Could not scroll to page. Reason: androidx.test.espresso.PerformException: Error performing 'ViewPager move to last page' on view 'Looked for element with View com.google.android.material.textview.MaterialTextView{dc69d1b VFED..C.. ........ 84,2995-1356,3177 #7f0a0074 app:id/btn_next}'.

Tried TouchActions also, did not work.

**Reply by KazuCocoa** [reply]
> Original error: androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: is assignable from class: class android.widget.AdapterView

Probably you should optimize the query for the adapterview... The message is raised when Espresso framework could not find an element by the given query. e.g. https://github.com/appium/appium-espresso-driver/blob/21ae67c6ae67d78aa068cc2c45d75c06015bc7cd/test/functional/commands/find-e2e-specs.js#L113

> Could not scroll to page. Reason: androidx.test.espresso.PerformException ...

Maybe https://developer.android.com/reference/androidx/test/espresso/contrib/ViewPagerActions did not work.
then, I guess you need to scroll with W3C actions, and check if an element exists by callingfind_elelemt.


> Tried TouchActions also, did not work.

Do you mean the TouchActions did not scroll a view by coordinate?

**Reply by attrideepak** [reply]
TouchActions did not scroll it. Here is snapshot of the U/I.
https://gist.github.com/attrideepak/2eb3205036de83ea92232f1db86a0205
The scenario is there is scroll view, inside the scroll view there is a button Next, which becomes visible on the screen after doing the scroll.
Also I am using this to perform the swipe, 
http://appium.io/docs/en/writing-running-appium/tutorial/swipe/simple-screen/

**Reply by attrideepak** [reply]
I am able to scroll using mobile command
```
 HashMap<String, String> scrollObject = new HashMap<>();
        scrollObject.put("element", scrollView.getId());
        scrollObject.put("direction", "up");
        driver.executeScript("mobile: swipe", scrollObject);
```
Do let me know if there is a better solution?

**Reply by KazuCocoa** [reply]
Did you try W3C actions? https://appium.io/docs/en/commands/interactions/actions/ (This is not https://appium.io/docs/en/writing-running-appium/touch-actions/)

Last one thing I could suggest is https://github.com/appium/appium-espresso-driver#mobile-swipe. `mobile: scrollToPage` and this method is via Espresso framework's API. If they also do not work, I guess Espresso framework by Android itself has an issue to interact with the scroll view.

btw, I experienced uiautomator by Android works but espresso by Android does not long before, even not via appium.
https://discuss.appium.io/  may help if someone has the same experience.

**Reply by attrideepak** [reply]
Found two solution for the problem:
W3C actions
`mobile:swipe`

**Reply by attrideepak** [reply]
thanks for the support
